### PR TITLE
Updated CircleCI config file to 2.0 

### LINKED ISSUE #1699: Update CircleCI config to 2.0
CircleCI, which we use to test our docker builds, has updated its config spec to 2.0. We should update our circleci config to 2.0, following [this guide](https://circleci.com/docs/2.0/migrating-from-1-2/) .

### PR DESCRIPTION
**Pull request by glenak1911**
Closes #1699

I've updated the config file according to [the documentation](https://circleci.com/docs/2.0/migrating-from-1-2/). I also had to consult a few other examples to try and get the syntax/semantics correct. 

I'd like to test this, but I'm not sure how to go about doing that(don't want to break any builds), as well as have this reviewed for accuracy. Any pointers would be helpful.

### CODE REVIEW

**Comment by minrk** [comment]
If you'd like to test, you can try enabling circleci on your own fork and push these same changes to your master branch. The docker login / upload step won't succeed because it won't have the credentials, but it should at least get that far.

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Python CircleCI 2.0 configuration file
+# Updating CircleCI configuration from v1 to v2
+# Check https://circleci.com/docs/2.0/language-python/ for more details
+#
+version: 2
+jobs:
+
+  test:
```
Comment (minrk) [inline .circleci/config.yml]: I was wrong. We do need a build job here. Or we need to use a custom [workflow](https://circleci.com/docs/2.0/workflows/).

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+              docker login -u $DOCKER_USER -p $DOCKER_PASS -e unused@example.com
+              docker push jupyterhub/jupyterhub-onbuild
+    release:
+      filters:
+        tags:
+          only: /.*/
+        steps:
+          - checkout
+          - run:
+            name: Push tagged branch to corresponding image
```
Comment (minrk) [inline .circleci/config.yml]: name and command here need to be indented one level beyond `run`:

```yaml
- run:
    name: ...
```

File: .circleci/config.yml (hunk 1 above)
Comment (minrk) [inline .circleci/config.yml]: I think all of the jobs need `machine: true` in the new 2.0 config

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+          command: |
+            docker build -t jupyterhub/jupyterhub .
+            docker build -t jupyterhub/jupyterhub-onbuild:${CIRCLE_TAG:-latest} onbuild
+
+  deploy:
+    hub:
+      branches:
+        only: master
+      steps:
+        - checkout
```
Comment (minrk) [inline .circleci/config.yml]: I don't think the deployments can have a checkout step

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  test:
+    steps:
+      - run:
+          name: Test image build
+          command: |
+            docker build -t jupyterhub/jupyterhub .
+            docker build -t jupyterhub/jupyterhub-onbuild:${CIRCLE_TAG:-latest} onbuild
+
+  deploy:
+    hub:
```
Comment (minrk) [inline .circleci/config.yml]: Deploy can't have sub-keys anymore. It can only have `steps`

**Comment by minrk** [comment]
I think we can actually remove the deploy and release steps and move those over to docker's own automated builds, if you'd like to try that. That should simplify things because then we only have the builds and no conditional login/push steps.

**Comment by minrk** [comment]
Ignore that last comment, I forgot about jupyterhub-onbuild, which would be a little tricky on docker. Let's try to get this working as-is here.

**Comment by glenak1911** [comment]
@minrk thanks for the reviews. I've made some changes, including adding workflows. It's still failing on my fork with the following:

`Configuration errors: 1 error occurred:

* Error migrating config to version 2: 2 errors occurred:

* in job 'workflows': steps is not a list
* in job 'workflows': steps is not a list`

The errors aren't very verbose, and I'm not sure if there is anything missing that is readily apparent to you.

**Comment by glenak1911** [comment]
I have to update my branch on my fork with the changes I made. Will post that later this evening.

File: .circleci/config.yml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      enabled: true
+    steps:
+      - checkout
+      - run:
+          command: docker login -u $DOCKER_USER -p $DOCKER_PASS -e unused@example.com
+      - run:
+          command: docker push jupyterhub/jupyterhub-onbuild:$CIRCLE_TAG
+
+
+  workflows:
```
Comment (minrk) [inline .circleci/config.yml]: workflows should be dedented so it aligns with 'jobs' rather than as an item under 'jobs'

**Comment by glenak1911** [comment]
@minrk Thanks for all the help! I tested this out in my fork and it looks good! I think this is complete.

File: package-lock.json [hunk 6]
```diff
@@ -0,0 +1,671 @@
+{
```
Comment (minrk) [inline package-lock.json]: Oops! This file shouldn't be committed.

Comment (minrk) [reply package-lock.json]: We should add it to .gitignore. Feel free to do that here if you want.

File: .circleci/config.yml [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+      - run:
+          command: docker build -t jupyterhub/jupyterhub .
+      - run:
+          command: docker build -t jupyterhub/jupyterhub-onbuild:${CIRCLE_TAG:-latest} onbuild
+
+  deploy:
+    machine:
+      enabled: true
+    steps:
+      - checkout
```
Comment (minrk) [inline .circleci/config.yml]: deploy and release shouldn't need checkout steps, only test

File: package-lock.json (hunk 6 above)
Comment (glenak1911) [reply package-lock.json]: ah understood, I've added updated the gitignore

File: .circleci/config.yml (hunk 7 above)
Comment (glenak1911) [reply .circleci/config.yml]: sounds good, I've removed the checkout step from deploy and release

**Comment by minrk** [comment]
Thanks!

**Pull request merged**